docs/MapHashers.md

```diff
@@ -0,0 +1,161 @@
+for test trees, where:
+ - The hash of a leaf is `HashLeaf(leaf) := HASH(0x00 || leaf.LeafValue)` for an arbitrary hash
+   function (SHA-256 by default).
+ - The hash of a non-empty interior node is `HashChildren(l, r) := HASH(0x01 || l || r)`.
+ - The hash of an empty subtree at level `n` is recursively build from the leaf
+   hashes of an empty leaf:
+     - `n = 0`: `E0 := HashLeaf(nil)`
+     - `n = 1`: `E1 := HashChildren(E0, E0)`
+     - `n = 2`: `E2 := HashChildren(E1, E1)`
+     - &hellip;
+
+This hash strategy has some simple properties:
+ - The hash of any node is independent of its address/location in the tree
+ - The hash of a zero-length leaf value is the same as the hash of an
+   empty (never set) leaf (i.e. `HashLeaf(nil) == HashEmpty(_, 0)`).
+ - More generally, the hash of the top of an empty subtree (whose leaves are all
+   empty/never-set) is the same as the hash of the same size subtree with leaves
+   with zero-length values.
+
+However, this means that this hash strategy is more vulnerable to attacks where
+subtrees are transplanted between different locations in the tree.
+
+### CONIKS Hasher Strategy
+
+The CONIKS hashing strategy
+([original paper](https://www.usenix.org/system/files/conference/usenixsecurity15/sec15-paper-melara.pdf))
+is more secure, but is more complicated because:
+ - Hash values always depend on the location / address of the node in the tree
+ - Empty subtrees have a different hash value than the equivalent subtree with
+   leaves that have zero-length values.  This has to be the case to preserve the
+   efficiency of the sparse Merkle tree: given that all hashes are location
+   dependent, if the hash of an empty subtree were the same as that of a
+   zero-length-leaves subtree, the hasher would need to calculate all
+   2<sup>257</sup> hashes for the whole Merkle tree.
+
+The details of the hash strategy are that:
+ - The hash of a leaf that exists is `HASH('L' || treeID || index || depth ||
+   leaf.LeafValue)`.
+ - The hash of a non-empty interior node is `HASH(l || r)`.
+ - The hash of an empty node is `HASH('E' || treeID || index || depth)`
+
+## Trillian Map API
+
+As described above, the map hashing strategy includes a specific hash calculation
+that applies to subtrees where all of the enclosed leaves are empty (have never
+been set).  This allows hash calculations to be more efficient for a sparse
+tree, but also allows for a more efficient Map API.
+
+In particular, any values in a leaf inclusion proof that match the empty hash
+value for that node need not be sent on the wire.  Instead, a `nil` value is used
+to indicate that this node in the proof is the top of an empty subtree (a
+subtree all of whose leaves are empty / never-been-set).
+
+Also, the distinction between empty / never-set leaves, and leaves whose values
+have been explicitly set to a zero-length value, means that two possible hash
+values apply for leaves with `len(leaf.LeafValue)==0`.
```

Comment:
Which can be distinguished with the value of `leaf.LeafHash`.  `len(leaf.LeafHash)==0` means that the value is empty. `len(leaf.LeafHash)!=0` means that `leaf.LeafValue` has been committed to explicitly, even if  `len(leaf.LeafValue) == 0`.

Reply:
Tweaked.